Incorporate duplicate into guides (#15)
ome/omero-guide-introduction: docs/data-management.rst

       |image2d|

.. warning::

>     **The ``Copy Link`` feature will not create a new independent copy of the Images.** The only way to create a fully independent copy in OMERO is to use the :ref:`Duplicate feature<Duplicatecli>`. The only thing that is copied during the actions above is the link between the Images and the Dataset. A single Image becomes linked to two Datasets.

Reviewer: ```suggestion
    **The ``Copy Link`` feature will only create new links between an Image and a Dataset, so that one Image becomes linked to multiple Datasets. This does not create a new independent copy of the Image.** The only way to create a fully independent copy in OMERO is to use the :ref:`Duplicate feature<Duplicatecli>`.
```
Note also that having shorter lines would be useful for review.

>     **The ``Copy Link`` feature will not create a new independent copy of the Images.** The only possiblity how to create a fully independent copy in OMERO is to use the :ref:`Duplicate feature<Duplicatecli>`. The only thing that is copied during the actions above is the link between the Images and the Dataset. A single Image becomes linked to two Datasets.

Reviewer 2: ```suggestion
    **The ``Copy Link`` feature will not create a new independent copy of the Images.** The only way to create a fully independent copy in OMERO is to use the :ref:`Duplicate feature<Duplicatecli>`. The only thing that is copied during the actions above is the link between the Images and the Dataset. A single Image becomes linked to two Datasets.
```

#. When creating new Datasets or Projects during the move, note that these containers will belong to the owner of the data, not yourself. Also the links between the new containers and the moved data will belong to the owner of the data. This should help to facilitate a smooth workflow, retaining the data handling possibilities such as reorganizing the data, renaming the containers you created for them etc. for the owner of the data. 

.. _Duplicatecli:

*Command Line: Duplicating objects*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You can duplicate objects in OMERO. The functionality is available only on the CLI for now. The duplication creates a fully independent duplicate of the duplicated object and, when desired, the objects linked to the duplicated object as well. When the duplicate is later deleted, it will not have an influence on the original, which stays preserved. Similarly, when the original is later deleted, the duplicate stays preserved as well.

Reviewer: ```suggestion
You can duplicate objects in OMERO. The functionality is available only on the CLI for now. The duplication creates a full copy of the objects as if they were created independently. When desired, the newly duplicated objects can additionally be linked to other objects. When the duplicate is later deleted, it will not have an influence on the original, which stays preserved. Similarly, when the original is later deleted, the duplicate stays preserved as well.
```

In case of Image objects, which have image files linked, the duplication creates a new image file which is linked to the original image file by a hard link. This means every duplication of an Image increases the number of hard links on the image file in Managed Repository of OMERO, but does not duplicate the image file itself, and thus does not increase the storage demands too much. 

> In case of File Attachment objects though, which also have files to them, each duplication will duplicate the linked file, thus doubling the storage space necessary for these File Attachment files.

Reviewer: Is this always the case? I thought a hard-link would be attempted first?

Reviewer 2: Not for attachments because of the risk of a user copying someone else's then writing to it.

Author: This sounds like the formulation is acceptable (let me know if I get this wrong)

Reviewer: I definitely stand (or sit) corrected, but I worry about the implications of this without control or feedback on how much space can be consumed.

Reviewer 2: Certainly not a big step from here to add some space limit.

Author: I guess this again shows that we should get this doc in fairly quickly to be able to pass such granular info to the user

Reviewer: :+1:

Resources
---------

-  Necessary software versions:

   - **OMERO.server 5.6.3 or later**
   - **omero-cli-duplicate plugin 0.4.0 or later**

-  Documentation:

   - `README of the omero-cli-duplicate repository <https://github.com/ome/omero-cli-duplicate/blob/master/README.rst>`_

   - `Application Programming Interface documentation <https://docs.openmicroscopy.org/omero-blitz/5.5.7/slice2html/omero/cmd/Duplicate.html>`_

-  Plugin for duplication on Command Line:

   - `omero-cli-duplicate on PyPI <https://pypi.org/project/omero-cli-duplicate/#description>`_

Setup
-----

**Duplicate plugin installation**

- Go to the environment where you installed your OMERO.cli as specified under -  https://docs.openmicroscopy.org/latest/omero/users/cli/installation.html.

- Activate the virtual environment.

- Run::
    
   $ pip install omero-cli-duplicate

Step-by-Step
------------

#. On your local machine, open a terminal

#. Activate the virtual environment where ``omero-py`` is installed or add it to ``PATH`` e.g.::
    

>    $ export PATH=/opt/omero/server/venv3/bin:$PATH

Reviewer 3: Seems that the more extensive "Activate..." instructions should be in the Setup section, then in the Step-by-Step you can just have "Activate the virtual environment." since you're just repeating what you've already explained how to do?

#. The variables ``$ID​1`` and ``$ID2`` below are the IDs of the ​selected Datasets. To duplicate two Datasets with their Images and annotations on both the Images and the Datasets, run::
    
   $ ​omero duplicate Dataset:$ID1,$ID2 --report

#. The duplicated Datasets will not be linked to any Project, even if the originals were linked to some Project. 

#. Duplicate two Images with many ROIs on them. The ROIs duplication might take a long time. To exclude the duplication of the ROIs, run::

   $ omero duplicate Image:$ID1,$ID2 --ignore-classes=Roi --report 

#. Find the duplicated Images in the Orphaned Images and Drag and Drop them into a Dataset or create a new Dataset for them.

> #. Dupliacte two Projects of another user in read-annotate group type. The group name in our example below is `read-annotate-group`. Log in as `user-1` and duplicate Projects belonging to `user-2` in that group. Further, instead of duplicating all the annotations of `user-2`, link all the original annotations such as Tags, Key-Value pairs and FileAnnotations of `user-2` to the corresponding objects in the duplicate, and duplicate only Comments and Ratings. The duplicate and Comments and Ratings on it will belong to `user-1`, but the other annotations will still belong to `user-2`. Run::

Reviewer 2: ```suggestion
#. Duplicate two Projects of another user in read-annotate group type. The group name in our example below is `read-annotate-group`. Log in as `user-1` and duplicate Projects belonging to `user-2` in that group. Further, instead of duplicating all the annotations of `user-2`, link all the original annotations such as Tags, Key-Value pairs and FileAnnotations of `user-2` to the corresponding objects in the duplicate, and duplicate only Comments and Ratings. The duplicate and Comments and Ratings on it will belong to `user-1`, but the other annotations will still belong to `user-2`. Run::
```

> #. Duplicate two Projects of another user in read-annotate group type. The group name in our example below is `read-annotate-group`. Log in as `user-1` and duplicate Projects belonging to `user-2` in that group. Further, instead of duplicating all the annotations of `user-2`, link all the original annotations such as Tags, Key-Value pairs and FileAnnotations of `user-2` to the corresponding objects in the duplicate, and duplicate only Comments and Ratings. The duplicate and Comments and Ratings on it will belong to `user-1`, but the other annotations will still belong to `user-2`. Run::

Reviewer 3: Suggested re-write (hopefully not too verbose)?

Duplicate two Projects of another user in read-annotate group type. Here we specify some `reference-classes` of objects that we do not want to duplicate, even though they are linked to the objects we are duplicating. Instead, the duplicated objects will be linked to these 'reference-classes' of object.
We can also specify classes that we explicitly do want to duplicate using the `duplicate-classes` argument.
In this example, we specify all Annotations as `reference-classes`, but a subset of these (Comments and LongAnnotations such as ratings) are `duplicate-classes`. This means that the duplicated Projects will be linked to the original annotations such as Tags, Key-Value pairs and FileAnnotations, but Comments and Ratings will be duplicated. This can be useful to ensure that a single Comment is not attached to multiple objects which might cause confusion when the comment is edited. The duplicated annotations will belong to the current user (user-1), even if all the original annotations may belong to other users in the read-annotate group. Run::

Author: > even if all the original annotations may belong to other users in the read-annotate group.

Yes, thanks, this is more precise. I did not think through the mixed annotation ownership on the original.

   $ omero logout
   $ omero login -u user-1 -g read-annotate-group
   $ omero duplicate Project:$ID1,$ID2 --reference-classes=Annotation --duplicate-classes=CommentAnnotation,LongAnnotation --report

> #. Dupliacte two Projects of another user in read-only group type. The group name in our example below is `read-only-group`. If the duplicator is not an administrator, administraotr with restricted privileges or group owner, they cannot link the annotations of another user to their duplicate in read-only group. They might duplicate all the annotations or exclude the duplication of all the annotations by excluding the Link duplication to the relevant objects as shown below. Run::

Reviewer 2: ```suggestion
#. Duplicate two Projects of another user in read-only group type. The group name in our example below is `read-only-group`. If the duplicator is not an administrator, administrator with restricted privileges or group owner, they cannot link the annotations of another user to their duplicate in a read-only group. They might duplicate all the annotations or exclude the duplication of all the annotations by excluding the Link duplication to the relevant objects as shown below. Run::
```

> 

Reviewer 3: Last sentence, I assume the default behaviour is to duplicate all annotations? Maybe just add that like: "They might duplicate all the annotations (this is the default behaviour) or ..."

Author: >  the default behaviour is to duplicate all annotations? 

Yes, it is afaik

   $ omero logout
   $ omero login -u user-1 -g read-only-group

>    $ omero duplicate Dataset:$ID1,$ID2 --ignore-classes=DatasetAnnotationLink,ImageAnnotationLink,Roi --report

Reviewer 2: ```suggestion
   $ omero duplicate Dataset:$ID1,$ID2 --ignore-classes=IAnnotationLink,Roi --report
```
probably also works if that's more attractive?

Author: It is shorter, but much harder to understand. I actually do not understand whether the result of the duplication will be exaclty the same as in the ``classes=DatasetAnnotationLink,ImageAnnotationLink`` or is it a slightly different behaviour (and result) which will also work (= will not error) which you are suggesting here ?

Author: Looking again on what the text above is saying, I think I understand now, it is readable and simpler like you suggest, thanks.
